pkg/localstorev2/internal/pinning/pinning.go

```diff
@@ -0,0 +1,281 @@
+)
+
+// batchSize used for deletion
+var batchSize = 100
+
+// CollectionStat is used to store some basic stats about the pinning collection
+type CollectionStat struct {
+	Total           uint64
+	DupInCollection uint64
+}
+
+// pinCollectionSize represents the size of the pinCollectionItem
+const pinCollectionItemSize = swarm.HashSize + uuidSize + 8 + 8
+
+// pinCollectionItem is the index used to describe a pinning collection. The Addr
+// is the root reference of the collection and UUID is a unique UUID for this collection.
+type pinCollectionItem struct {
+	Addr swarm.Address
+	UUID string
+	Stat CollectionStat
+}
+
+// common prefix for all pinCollectionItem entries
+func (pinCollectionItem) Namespace() string { return "pinCollectionItem" }
+
+func (p *pinCollectionItem) ID() string { return p.Addr.ByteString() }
+
+func (p *pinCollectionItem) Marshal() ([]byte, error) {
+	if p.Addr.IsZero() {
+		return nil, errInvalidPinCollectionAddr
+	}
+	if len(p.UUID) == 0 {
+		return nil, errInvalidPinCollectionUUID
+	}
+	buf := p.Addr.Bytes()
+	buf = append(buf, []byte(p.UUID)...)
+	statBuf := make([]byte, 8+8)
+	binary.LittleEndian.PutUint64(statBuf[:8], p.Stat.Total)
+	binary.LittleEndian.PutUint64(statBuf[8:16], p.Stat.DupInCollection)
+	buf = append(buf, statBuf...)
+	return buf, nil
+}
+
+func (p *pinCollectionItem) Unmarshal(buf []byte) error {
+	if len(buf) != pinCollectionItemSize {
+		return errInvalidPinCollectionSize
+	}
+	ni := new(pinCollectionItem)
+	ni.Addr = swarm.NewAddress(buf[:swarm.HashSize])
+	ni.UUID = string(buf[swarm.HashSize : swarm.HashSize+uuidSize])
+	statBuf := buf[swarm.HashSize+uuidSize:]
+	ni.Stat.Total = binary.LittleEndian.Uint64(statBuf[:8])
+	ni.Stat.DupInCollection = binary.LittleEndian.Uint64(statBuf[8:16])
+	*p = *ni
+	return nil
+}
+
+// pinChunkItem is the index used to represent a single chunk in the pinning
+// collection. It is prefixed with the UUID of the collection.
+type pinChunkItem struct {
+	UUID string
+	Addr swarm.Address
+}
+
+func (p *pinChunkItem) Namespace() string { return p.UUID }
+
+func (p *pinChunkItem) ID() string { return p.Addr.ByteString() }
+
+// pinChunkItem is a key-only type index. We don't need to store any value. As such
+// the serialization functions would be no-ops. A Get operation on this key is not
+// required as the key would constitute the item. Usually these type of indexes are
+// useful for key-only iterations.
+func (p *pinChunkItem) Marshal() ([]byte, error) {
+	return []byte{}, nil
+}
+
+func (p *pinChunkItem) Unmarshal(_ []byte) error {
+	return nil
+}
+
+// PutterCloser implements the storage.Putter as well as an io.Closer. The Putter
+// should be closed for a successful operation.
+type PutterCloser interface {
+	storage.Putter
+	io.Closer
+}
+
+// Storage interface is the functionality required from the underlying storage to
+// achieve the pinstore functionality. This allows us to do stateless operations
+// here.
+type Storage interface {
```

Comment:
I think this should be moved to the `storage.go` file, or?

Reply:
So I wrote some comments on your PR as well. I dont want to move this to `storagev2` pkg as these could be specific combinations of some interfaces in the `storage` pkg. Once we have all the components, if these are common for all, maybe it would make sense to have this there.

The way I am thinking about this is the internal packages could have some interfaces defined, and they could be local to the internal pkgs and the transactions would basically have to implement this. So I would keep this here for now.